# multiomics_MDD

<!-- [![DOI](https://zenodo.org/ lien à ajouter -->

This repository houses the code used in our study of multiomic peripheral blood biomarkers for Major Depressive Disorder (MDD). We integrated gene expression data with DNA methylation (DNAm) and microRNAs (miRNAs) derived from peripheral blood samples from a well-characterized cohort of individuals with MDD (n=80) and healthy controls (n=89). The raw and processed data are available on GEO with accession number [GSE251786](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE251786) and will be made publically available at the paper acceptance. 

The repository **multiomics_MDD** is divided into three main parts: 
- [1- Single omics](https://github.com/INSERM-U1141-Neurodiderot/multiomics_MDD/tree/main/1_Single_Omics)
- [2- RRHO2](https://github.com/INSERM-U1141-Neurodiderot/multiomics_MDD/tree/main/2_RRHO2)
- [3- Multiomic workflow](https://github.com/INSERM-U1141-Neurodiderot/multiomics_MDD/tree/main/3_multiomics)

## 1- Single omics

We first identified DMPs, DEmiRNAs and DEGs between MDD cases and controls. The code is presented in the Jupyter Notebook [1_Single_Omics/README.md](https://github.com/INSERM-U1141-Neurodiderot/multiomics_MDD/blob/main/1_Single_Omics/README.md) listing R packages and fonctions used. 

We also provided the processed data and covariates used as input:
- DNAm processed data ([data_DNAm.rds](https://github.com/INSERM-U1141-Neurodiderot/multiomics_MDD/blob/main/1_Single_Omics/data.mRNA.rds))
- DNAm covariates ([LeucocyteFraction.mdd.RDS](https://github.com/INSERM-U1141-Neurodiderot/multiomics_MDD/blob/main/1_Single_Omics/LeucocyteFraction.mdd.RDS))
- mRNA processed data ([data.mRNA.rds](https://github.com/INSERM-U1141-Neurodiderot/multiomics_MDD/blob/main/1_Single_Omics/data.mRNA.rds))
- miRNA covariates ([miR.MDD.raw.renamed.counts.csv](https://github.com/INSERM-U1141-Neurodiderot/multiomics_MDD/blob/main/1_Single_Omics/miR.MDD.raw.renamed.counts.csv))
- miRNA processed data ([covMIRNA.mDD.RDS](https://github.com/INSERM-U1141-Neurodiderot/multiomics_MDD/blob/main/1_Single_Omics/covMIRNA.mDD.RDS))

## 2- RRHO2

We implemented **RRHO2par** a parallelized version of **RRHO2** ([Cahill et al. 2018](https://doi.org/10.1038/s41598-018-27903-2)). The code is presented in the Jupyter Notebook [2_RRHO2/Readme.md](https://github.com/INSERM-U1141-Neurodiderot/multiomics_MDD/blob/main/2_RRHO2/Readme.md). We applied **RRHO2par** to the R data object *res_single_mRNA.rds* generated by the mRNA single omic analysis (cf. line 187 [1_Single_Omics/README.md](https://github.com/INSERM-U1141-Neurodiderot/multiomics_MDD/blob/main/1_Single_Omics/README.md))
 
## 3- Multiomic workflow
Finally, with the goal of fully leveraging our genome-wide and multiomic data, we implemented a supervised integration framework designed to identify features that best discriminate MDD patients from healthy controls. Our framework was developed, building on the [Momix package](https://github.com/cantinilab/momix-notebook) ([Cantini et al. 2021](https://doi.org/10.1038/s41467-020-20430-7)) and Similarity Network Fusion (SNF, [Wang et al. 2014](http://www.nature.com/nmeth/journal/v11/n3/full/nmeth.2810.html)), both encapsulated in a Cross-Validation (CV) procedure. The code is presented in several R scripts :

.... à revoir :

La partie Snake ajoutée et commentée par Claire-Cécile pour l'instant ne représente qu'une toute petite partie du framework (Momix essentiellement) à retirer du github pour le papier

Pour l'instant l'ordre des script tel que présenté ci-dessous n'est pas bon avec les scripts actuels, mais collerait avec la presentation dans le Mat&Met --> à adapter aux scripts finalisés

Il y a du preproccessing et de la cross-validation dans le script SNFclustering.R --> à bouger ou unifier? 
(embedded CV for SNF optimization ? normalement non) 

Le script SNFclustering.R contient des ARI et nemo à supprimer? a priori on a besoin que des AUC et de SNF

Pour présenter SNFclustering, si j'ai bien compris, Arnaud proposait de simplifier en montrant seulement un example pour une "split-itération" de CV dans une combinaison omic-cohort pour une jDR méthode (est-ce que ce serait faisable pour combinaison RNAseq miR DNAm JIVE female par ex?) 

Le script de bootstrapping semble être pour JIVE. Est-ce que vous pourriez le vérifier et le commenter?

Pour les fichiers de données de la partie multiomique du GitHub, ne garder que les inputs et outputs du script example? 

- Pre-processing and Cross-Validations : ([Cross_validation.R](https://github.com/INSERM-U1141-Neurodiderot/multiomics_MDD/blob/main/3_multiomics/Cross_validation.R))
- Feature selection with joint Dimension Reduction (jDR) methods ([Variable_selection.R](https://github.com/INSERM-U1141-Neurodiderot/multiomics_MDD/blob/main/3_multiomics/Variable_selection.R))
- Multiomic clustering with SNF: ([SNF_clustering.R](https://github.com/INSERM-U1141-Neurodiderot/multiomics_MDD/blob/main/3_multiomics/SNF_clustering.R))
- Evaluation of the predicting performance for each combination of omic and cohort ([Results.R](https://github.com/INSERM-U1141-Neurodiderot/multiomics_MDD/blob/main/3_multiomics/Results.R))
- Bootstrapping procedure to extract stable selected features: Example for JIVE ([Jive_boost.R](https://github.com/INSERM-U1141-Neurodiderot/multiomics_MDD/blob/main/3_multiomics/Jive_boost.R))


## Cite

To cite this work please use the reference of the following paper:

+ Formatted citation:
  
> Mokhtari et al., (2023). Sex-specific and multiomic integration enhance accuracy of peripheral blood biomarkers of major depressive disorder <!--lien research square à ajouter https://doi -->

<!-- + BibTeX citation: -->

